### DISCUSSION TITLE
Add Chatterino as a Dock in OBS

### DISCUSSION THREAD

**Discussion by Nerixyz** [Chatterino/chatterino2#4669] [Ideas]
In [OBS](https://obsproject.com/) you can customize the UI for your liking by moving parts of the UI - docks ([`QDockWidget`](https://doc.qt.io/qt-6/qdockwidget.html)s) - around. You can add custom browser docks which load a website. Some streamers might want to add a Twitch chat into their OBS. But since this is essentially a web-view, the possibilities are limited. However, you can also add custom docks through plugins ([`obs_frontend_add_dock`](https://docs.obsproject.com/reference-frontend-api#c.obs_frontend_add_dock)). So one might want to add Chatterino as a dock which would/could provide a better chat experience as well as e.g. user cards that aren't possible in the browser widget.

Since Chatterino runs on Qt (and supports Qt 6), Chatterino could be added as a dock. I'd imagine that this would involve substantial changes to the build process. The static `chatterino-lib` might need to become a dynamic library that could be loaded _by_ an OBS plugin, which would then create the dock - this plugin would be a separate project.

**Comment by Nerixyz** [comment]
Some thoughts since this was asked about in the Discord:

- Splitting Chatterino into a dynamic library and an executable might not be a good idea. The main reason for this is that we'll need to use a binary compatible Qt version for users to be able to copy-paste the library to the OBS plugins.
- The better approach would be to compile the plugin separately to the Qt version of OBS and statically link to Chatterino (`chatterino-lib`). There might be some issues with network dependencies, but that would only require OpenSSL to be copied.
- Most widgets in Chatterino are (sometimes implicitly) dependent on others both up and down the tree. Specifically, some widgets might try to access `Notebook`. In an OBS plugin, I would expect us to show one split per pane, so there wouldn't be any notebook.
- Global state might be a bit awkward to work with in a plugin. We have to ensure that we don't crash the host.
- How do we handle settings in the plugin? On one hand, you'd want the plugin to pick up the global account settings, but on the other hand, you'd want to have some "overlay" settings in the plugin (maybe a different zoom level). Furthermore, you wouldn't want the plugin to mess with the settings of the standalone app.

Note that not all of these have to be resolved in some initial version (e.g. the initial version could pick up global settings). We can work incrementally (and I'd prefer to do so).

**Comment by Nerixyz** [comment]
I've been hacking around on this a bit and created https://github.com/Nerixyz/chatterino-obs. You can find prebuilt versions here: https://github.com/Nerixyz/chatterino-obs/releases.
> [!WARNING]
> These are development versions. While they may work today, they might break tomorrow, and crash your OBS while you're looking away.

https://github.com/user-attachments/assets/d5c1b535-2e7c-40af-82cd-e4341b92a676

As you can hopefully tell by the readme, building this isn't fun. That's mainly because we're forced to use MSBuild on Windows and Xcode on macOS as the CMake generator. Neither of these allows generating a `compile_commands.json` 😦 And for some reason, Xcode doesn't build some files, so I disabled macOS builds. We also need to ship TLS plugins for Qt that aren't included with OBS. Unfortunately, the prebuilt Qt from OBS doesn't build the OpenSSL plugin, so the system TLS library will be used. These have caused issues on some systems in the past.

I'm using a fork of Chatterino (https://github.com/Nerixyz/chatterino2/tree/feat/chatterino-obs) with some patches to be able to use it with `add_submodule`.

To help with the build problems, I had the idea of building a shared library of Chatterino that exposes the necessary functions to be able to embed it in another application (either in-tree or out-of-tree). This would resolve the problem of weird CMake generators, allow forks to ship their own embeddable library, and it could help with forwards compatibility (OBS updating their Qt without the user updating the plugin).

The actual OBS plugin would use the shared library and take care of initializing the app, creating UI, and saving dock positions.

If this sounds like a good approach, I'm also happy to move this into the Chatterino GitHub organization.

**Reply by pajlada** [reply]
Modifying our main repo CMake files to build a shared library that can be used by the OBS integration seems good to me, so overall I think this sounds like a good approach.
